include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                    copy.outMipLevel = 0u;
+                    copy.inImage = _img;
+                    copy.outImage = storage->image.get();
+                    copy.axisWraps[0] = _wrapu;
+                    copy.axisWraps[1] = _wrapv;
+                    copy.axisWraps[2] = ISampler::ETC_CLAMP_TO_EDGE;
+                    copy.borderColor = _borderColor;
+                    if (!CPaddedCopyImageFilter::execute(&copy))
+                        _IRR_DEBUG_BREAK_IF(true);
+                }
+        }
+
+        return offsetToTextureData(pgtOffset, _img, _wrapu, _wrapv);
+    }
+
+    bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return false;
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+        }
+
+        //free physical pages
+        auto extent = _img->getCreationParameters().extent;
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        CFillImageFilter::state_type fill;
+        fill.outImage = m_pageTable.get();
+        fill.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        fill.subresource.baseArrayLayer = _addr.pgTab_layer;
+        fill.subresource.layerCount = 1u;
+        fill.fillValue.asUint.x = SPhysPgOffset::invalid_addr;
+
+        using phys_pg_addr_alctr_t = ICPUVTResidentStorage::phys_pg_addr_alctr_t;
+
+        auto* const bufptr = reinterpret_cast<uint8_t*>(m_pageTable->getBuffer()->getPointer());
+        for (uint32_t i = 0u; i < levelCount; ++i)
+        {
+            const uint32_t w = neededPageCountForSide(extent.width, _subres.baseMipLevel+i);
+            const uint32_t h = neededPageCountForSide(extent.height, _subres.baseMipLevel+i);
+
+            const auto& region = m_pageTable->getRegions().begin()[i];
+            const auto strides = region.getByteStrides(TexelBlockInfo(m_pageTable->getCreationParameters().format));
+
+            for (uint32_t y = 0u; y < h; ++y)
+                for (uint32_t x = 0u; x < w; ++x)
+                {
+                    uint32_t* texelptr = reinterpret_cast<uint32_t*>(bufptr + region.getByteOffset(core::vector4du32_SIMD((_addr.pgTab_x>>i) + x, (_addr.pgTab_y>>i) + y, 0u, _addr.pgTab_layer), strides));
+                    SPhysPgOffset physPgOffset = *texelptr;
+                    if (storage->physPgOffset_valid(physPgOffset))
+                    {
+                        *texelptr = SPhysPgOffset::invalid_addr;
+
+                        uint32_t addrs[2] { physPgOffset.addr&0xffffu, storage->physPgOffset_mipTailAddr(physPgOffset).addr&0xffffu };
+                        const uint32_t szs[2]{ 1u,1u };
```

Comment:
because you know the max allocatable texture size in the constructor, you also know the max amount of pages (lets' call it `m_maxAllocationPageCount`) that could be allocated.

So you could preallocate `smart_refctd_dynamic_array`s for the temporary `addrs` and `szs` (`szs` contents would be constant, all 1s)

then loops like these would only fill out the `addrs` (if you fill `addrs`  entry with the special invalid value, then allocator  traits' multi_free will skip freeing it)

and then only fire off one `multi_alloc_addr` or `multi_free_addr`

Reply:
the need for this is very evident because you'll need to decouple the Page Table update loop from the Physical Storage allocator `multi_free_addr`